File: cpp/04-gtest-mock/test/CurrencyCalculatorTest.cpp
#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include "../src/CurrencyCalculator.hpp"
#include "../src/RateService.hpp"
#include "../src/Money.hpp"

// Mock class for RateService
class MockRateService : public RateService {
public:
    MOCK_METHOD(std::optional<double>, getRate, (const std::string& currency), (override));
};

class CurrencyCalculatorTest : public ::testing::Test {
protected:
    std::shared_ptr<MockRateService> mockRateService;
    std::shared_ptr<CurrencyCalculator> sut;

    void SetUp() override {
        mockRateService = std::make_shared<MockRateService>();
        sut = std::make_shared<CurrencyCalculator>(mockRateService);
    }
};

TEST_F(CurrencyCalculatorTest, testRunWithValidInput) {
    // given
    EXPECT_CALL(*mockRateService, getRate("CNY")).WillOnce(testing::Return(7.20));
    EXPECT_CALL(*mockRateService, getRate("EUR")).WillOnce(testing::Return(0.85));

    // when
    Money result = sut->convert(100.00, "CNY", "EUR");

    // then
    EXPECT_DOUBLE_EQ(11.81, result.getAmount());
    EXPECT_EQ("EUR", result.getCurrency());
}

File: java/04-with-mock/src/test/java/com/foo/currency/CurrencyCalculatorTest.java
package com.foo.currency;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.*;

/**
 * CurrencyMain 的单元测试，演示如何 mock 依赖组件
 */
@ExtendWith(MockitoExtension.class)
public class CurrencyCalculatorTest {
    
    @Mock//←这个是待测对象（sut）mockito框架会自己创建好
    private RateService mockRateService;

    @InjectMocks//←框架会自动创建类实例，并且注入依赖
    com.foo.currency.CurrencyCalculator sut;

    @Test
    public void testRunWithValidInput() throws Exception {
        // given
        when(mockRateService.getRate("CNY")).thenReturn(new BigDecimal("7.20"));
        when(mockRateService.getRate("EUR")).thenReturn(new BigDecimal("0.85"));

        // when
        Money result = sut.convert(BigDecimal.valueOf(100.00), "CNY", "EUR");
        // then
        assertEquals("11.81", result.getAmount().toString());
        assertEquals("EUR", result.getCurrency().toString());
    }
    
}
